### ISSUE TITLE
Polymer 3 support

### ISSUE THREAD

**Issue by BorntraegerMarc** [hotforfeature/origami#57]
Just wanted to ask if/how and when you think polymer 3 should be supported with origami? 😄 Maybe we could release a `-beta` version to let people play around with it?

**Comment by hotforfeature** [comment]
Since the Polymer 3 API isn't changing much at all, Origami should theoretically work with it out of the box. What we'd need to do is update the docs with a new build method.

This past week I've been working on a newer build system using `polymer-webpack-loader` by importing dependencies directly into Angular files.

my.component.ts
```ts
import { Component } from '@angular/core';

import 'assets/bower_components/paper-button/paper-button.html';

@Component({
  selector: 'my-component',
  template: '<paper-button>Hello from Angular</paper-button>'
})
export class MyComponent() { }
```

This has proven extremely useful in a project I'm working on. It gets rid of the `elements.html` bundled import, the entire Polymer CLI hacky build step, and allows Angular and webpack to do some optimization with chunks.

Downside is that right now I have to run a custom Angular CLI in order to get the loaders working, but it'd work with an ejected CLI or when the Angular team introduced hooks/customizations that they've been talking about on GitHub.

This partial transition to "sort-of-es-modules" will be a stepping stone to Polymer 3. Instead of importing the html with `polymer-webpack-loader`, you'd just import the module directly.

```ts
import { Component } from '@angular/core';

import '@polymer/paper-button/paper-button';

@Component({
  selector: 'my-component',
  template: '<paper-button>Hello from Angular</paper-button>'
})
export class MyComponent() { }
```

I'm hoping to refine this process in the next couple of weeks and update the docs and demo app to reflect it. This is my current thought and direction for Polymer 3 support though :)

**Comment by BorntraegerMarc** [comment]
Love it!!! 😄

**Comment by alexeygt** [comment]
Looking forward to try it! :+1:

**Comment by jagcolombat** [comment]
Hello  @hotforfeature, i am using this fantastic library in a evaluation project, and i have trying use it with Polymer 3 components, but with the described process in documentation no work.
I was make some modifications to the library and i have obtained sucessfull result. 
I will like share my experence for evaluation and perhaps contribute to this project. 
How i must do?

**Comment by hotforfeature** [comment]
@jagcolombat I'd recommend opening a PR with your changes. We can then review them and see what can be merged into a new branch for Polymer 3 support

**Comment by cdchow** [comment]
Hi @hotforfeature! Are there any updates on this? I'm using `origami` with Polymer 3 based components (https://vaadin.com/components/vaadin-text-field) in an Angular 5 project. I roughly followed your directions, with exceptions, and got it to work in local development:  
`ng serve`
<img width="279" alt="screen shot 2018-05-31 at 9 13 48 am" src="https://user-images.githubusercontent.com/21041308/40794051-f550fefa-64b2-11e8-9c88-130e28d64f00.png">

The exceptions being:
1. `webcomponentsjs/webcomponents-loader.js` was placed in the scripts in `.angular-cli.json` instead of the index.html
2. did not run/use the patch-cli because these Polymer elements are ES6 modules
3. when doing a production build (`ng serve -- prod` or `ng build --prod`) there client errors on the browser traced to uglified variables in `webcomponentsjs/shadycss/src/custom-style-interface.js:85` 
```
Uncaught ReferenceError: a20867 is not defined
    at d.getStyleForCustomStyle
``` 

relatedly, here are the relevant package and versions of my setup:
```
    "@angular/core": "^5.0.0",
    "@codebakery/origami": "^2.0.3",
    "@webcomponents/webcomponentsjs": "^2.0.0",
```

In summary, I was wondering if you could possibly provide some advice on #3 above. It appears that there could be a bug in AngularCLI's webpack uglifying a dependency ...

**Comment by hotforfeature** [comment]
Hey @cdchow! Thanks for your input and research. 

Can you tell what is importing shadycss from the source folder? The only element I know of that imports the CustomStyleInterface is `polymer/lib/elements/custom-style.js`, and it should be importing from `@webcomponents/shadycss/entrypoints/custom-style-interface.js` and not the src folder.

Additionally, can you check what the version of `node_modules/@webcomponents/shadycss/package.json` is?

Also, can you try running `ng serve --aot` and `ng serve --prod --build-optimizer=false`? The prod flag does quite a few things, so that'll help narrow it down if it runs correctly on either of those commands.

**Comment by cdchow** [comment]
Thanks for the quick response, it is thoroughly appreciated!

I didn't notice the `entrypoints` directory. I did `ng serve --prod --sourcemaps` to trace to the file that threw the error, wasn't sure what directory it was located.

For: `node_modules/@webcomponents/shadycss/package.json`
```
  "name": "@webcomponents/shadycss",
  "version": "1.2.1",
```

1) `ng serve --aot` is fine, I see the polymer component
2) `ng serve --prod --build-optimizer=false` doesn't work but gives slightly more informative error tracing
![screen shot 2018-05-31 at 11 41 50 am](https://user-images.githubusercontent.com/21041308/40801385-ae08e494-64c7-11e8-9848-cc9c47ec5f78.png)

**Comment by cdchow** [comment]
@hotforfeature ^

**Comment by hotforfeature** [comment]
@cdchow Do you have a repo that I could clone to take a look at?

**Comment by cdchow** [comment]
@hotforfeature I'm working on a private project and tried setting up a separate repo with the exact implementation I used, and looks like it's fine with both commands (with and without --prod). This means that something in my project is wrong...

I suspect updating my angular version might fix this as my project is `"@angular/core": "^5.0.0"` while this separate repo is using a generated `"@angular/core": "^5.2.0"`

**Comment by hotforfeature** [comment]
What's the `@angular/cli` version? You might try updating that first since it'll be whatever the CLI's doing with its build optimizer.

**Comment by cdchow** [comment]
I cannot reproduce those errors after updating! I updated a bunch of @angular dependencies to `"^5.2.0"`. I agree with your suggestion, but my private project @angular/cli version is actually higher than the same project. Now that the client errors no longer happen, I'm unblocked!

my private project:
```
"@angular/cli": "1.7.3",
"@angular/compiler-cli": "^5.0.0",
``

separate repo: https://github.com/cdchow/angular-origami-polymer-vaadin
```
"@angular/cli": "1.6.6",
"@angular/compiler-cli": "^5.2.0",
```

**Comment by jagcolombat** [comment]
Greetings @hotforfeature and @cdchow, i was reading your comments created yesterday and feel very happy for touch this theme again and for the test and configuration on **separate repo** of @cdchow. 

On April i wrote a article about this on https://medium.com/@tonytunes2005/integrating-polymer-3-components-on-angular-5-more-simple-7db2c5471567. But i saw in the **separate repo** of @cdchow that the webcomponents-loader.js was included on .angular-cli.json on script array and this is bit different to my described tests in the article, like me interesting that and decided proove this approach.

As said @cdchow it work!!!, but i try this approach in Firefox 59 and not work. I suggest you that if you expect your project work in this browser test it.

In other hand, i tested the second approach described on the article https://medium.com/@tonytunes2005/integrating-polymer-3-components-on-angular-5-more-simple-7db2c5471567#375c with the @webcomponents/webcomponentsjs@2.0.0 and luckily work, and now both approach pass the test!!! Later i´ll update de article.
 
Thanks for the contributions and motivation.

**Comment by cdchow** [comment]
My previous comment was somewhat misleading -- I had a stale yarn_cache and it somehow didn't update my dependencies to the latest minor version (via the ~ in the version definition). Pruning my yarn cache and getting latest minor versions is closer to the fix than updating to `~5.2.0`

**Comment by cdchow** [comment]
@hotforfeature are you looking for contributors? I still retained context from my endeavor and would be more than happy to incorporate a more well documented sample Angular5/Polymer3 example into your repo

**Comment by rjsteinert** [comment]
@hotforfeature From my experiments today with a fresh Angular CLI generated project, it seems like we might be close to not needing an additional dependency other than webcomponentjs added via `./src/polyfills.ts` to get Polymer working with Angular. When you get a chance to play around with it, I'm interested in what you find out.

**Comment by jagcolombat** [comment]
Hi @rjsteinert I would like know specifically that angular-cli version and browsers you used in yours experiments. Recently used angular-cli 6 and for the [(ngModel)] directive i needed ironControl directive. Certainly, I can use other way but ironControl continues to be  usefull. I would like share a link to one article related to one experiment with angular 6, polymer 3 and origami https://medium.com/@tonytunes2005/polymer-3-this-time-with-angular-5-6-e03f6bc0c027.

**Comment by hotforfeature** [comment]
Quick update for those who didn't see in Slack. Work has begun on a v3 rewrite that aims to be faster, smaller, and include Polymer 3 support. I've recruited @asyncLiz to help me with it, and we hope to have a new release in the next week or two!

**Comment by jorgecasar** [comment]
Any update? How can I install v3? I'm doing some tests with Angular 6 and Polymer 3

**Comment by asyncLiz** [comment]
Hey @jorgecasar! The bulk of the work is complete on the [v3 branch](https://github.com/hotforfeature/origami/tree/v3). The only thing left is to add tests, finish up some documentation, and provide better utility for setting up polyfills.

I'm hoping to be done with the documentation and final tweaks this week, but it may not be until next week that testing is complete and we're confident enough to cut a release. You can always drop by the #polymer-angular channel in https://polymer.slack.com/. I'm giving more frequent updates there 😄

**Comment by asyncLiz** [comment]
Polymer 3 is now supported with the v3 release!
